physical/raft/fsm.go

```diff
@@ -684,15 +697,18 @@ func (f *FSM) ApplyBatch(logs []*raft.Log) []interface{} {
 		f.applyCallback()
 	}
 
-	err = f.db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket(dataBucketName)
-		for _, commandRaw := range commands {
-			entrySlice := make([]*FSMEntry, 0)
+	for _, commandRaw := range commands {
+		inTx := false
+		entrySlice := make([]*FSMEntry, 0)
+		err = f.db.Update(func(tx *bolt.Tx) error {
+			b := tx.Bucket(dataBucketName)
 			switch command := commandRaw.(type) {
 			case *LogData:
 				for _, op := range command.Operations {
 					var err error
 					switch op.OpType {
+					case beginTxOp, commitTxOp:
```

Comment:
Probably worth sanity-checking that the first operation is a `beginTxOp` and the last is a `commitTxOp` and we never see these appear anywhere else.